src/access/wac.ts

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright 2020 Inrupt Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to use,
+ * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+ * Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { internal_fetchAcl } from "../acl/acl.internal";
+import { getAgentAccess as getAgentAccessWac } from "../acl/agent";
+import { WebId, WithServerResourceInfo } from "../interfaces";
+import { internal_defaultFetchOptions } from "../resource/resource";
+import { Access } from "./universal";
+import { Access as AclAccess } from "../acl/acl";
+
+function getDefaultAccess(): Access {
+  return {} as Access;
+}
+
+function aclAccessToUniversal(access: AclAccess): Access {
+  const universalAccess = getDefaultAccess();
+  // In ACL, denying access to an actor is a notion that doesn't exist, so an
+  // access is either granted or not for a given mode.
+  // This creates a misalignment with the ACP notion of an access being granted,
+  // denied, or simply not mentionned. Here, we convert the boolean vision of
+  // ACL into the boolean or undefined vision of ACP.
+  for (const [mode, accessMode] of Object.entries(access)) {
+    if (accessMode) {
+      // There is no convenient way to type `mode` as the keys to the Access interface
+      // @ts-ignore
+      universalAccess[mode] = accessMode;
+    }
+  }
+  return universalAccess;
+}
```

Comment:
I think I'd prefer to just explicitly set the relevant modes, so as to ensure that future changes in either interface will be caught here.

And actually, maybe we want to define an interface for the WAC-specific subsets of the possible Access Modes that would also be used as the return value for `getAgentAccess`, e.g. something like this:

```typescript
type WacAccess = Required<Access> & (
  { controlRead: true, controlWrite: true } |
  { controlRead: false, controlWrite: false }
);
```

(`Required<>` indicates that `undefined` is not an option, and the disjunction that `controlRead` and `controlWrite` will always have the same value.)

And so this would be something like:

```suggestion
function aclAccessToUniversal(access: AclAccess): WacAccess {
  return {
    read: access.read === true,
    append: access.append === true,
    write: access.write === true,
    controlRead: access.control === true,
    controlWrite: access.control === true,
  } as WacAccess;
}
```

This would also remove the need for the `getDefaultAccess` function.

Reply:
Oh hmm, in fact... Maybe `WacAccess` should have each mode be `undefined | true` instead? Since access that is not specified can still be granted via e.g. a group, so in that sense it's different from an actual Deny.

Reply:
Exactly, that's the intent behind this: in the universal API, access modes are aligned with the ACP logic, namely true if explicitly granted, false if explicitly denied, and `undefined` if neither. That said, creating an explicit `WacAccess` type and only setting the corresponding types in `aclAccessToUniversal` makes perfect sense.

Reply:
Right, since `WacAccess` is still just a subset of `Access`, i.e. `false` is not actually a valid value.

Reply:
280cf4d09004624662e0a440c419db280a687d7b